common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -778,11 +773,6 @@ public void onComplete(String streamId) throws IOException {
               // IOException to the client. This may increase the chunk size however the increase is
               // still limited because of the limit on the number of IOExceptions for a
               // particular shuffle partition.
-            } catch (NullPointerException e) {
-              throw new RuntimeException(
-                String.format("The merged shuffle partition info for appId %s shuffleId %s "
-                  + "reduceId %s has been cleaned up", partitionInfo.appId,
-                  partitionInfo.shuffleId, partitionInfo.reduceId));
```

Comment:
As discussed offline, let us not do this - this is inherently buggy approach.
Let us do the alternative discussed -

a) Remove direct field reference, and rely on accessor methods.
b) In the getter, do the relevant checks to ensure non-null values.
c) Use `AtomicReference` for the fields in partition info - given they are getting mutated async.
d) Ensure close/cleanup relies on atomic getAndSet to ensure no MT-safety issues.

Reply:
I agree that this way we'd avoid NPE reference but just wonder it can't tell the NPE reason explicitly. NPE can not only cause by `AppShufflePartitionInfo` cleanup but also by potential buggy code. In the way you proposed, I can't see how it could distinguish these two cases.

And, how about we add an atomic flag to indicate `AppShufflePartitionInfo` whether it's cleaned up or not. Then, when never we try to write the partition, we check the flag first.

Reply:
BTW, when do we clean up data/index/metadata files? I suppose they should also be clean up when `AppShufflePartitionInfo` is being clean up, shouldn't they?

Reply:
> BTW, when do we clean up data/index/metadata files? I suppose they should also be clean up when `AppShufflePartitionInfo` is being clean up, shouldn't they?

The `closeAndDeletePartitionFilesIfNeeded` will close the channels in the `AppShufflePartitionInfo` and delete those files if the deletion flag is set as True. If a newer attempt gets registered, the deletion flag will be set to true, so those files will be deleted. @Ngone51

Reply:
> I agree that this way we'd avoid NPE reference but just wonder it can't tell the NPE reason explicitly. NPE can not only cause by `AppShufflePartitionInfo` cleanup but also by potential buggy code. In the way you proposed, I can't see how it could distinguish these two cases.
> 
> How about we add an atomic flag to indicate `AppShufflePartitionInfo` whether it's cleaned up or not? Then, when never we try to write the partition, we'd check the flag first.

Setting a `AtomicBoolean` cleanup flag won't work and it would still throw NPE in some cases here. Suppose Thread 1 is doing the `AtomicBoolean.get`, and it returns false, which indicates the cleanup is not yet done, then it should continue to run the dataChannel.flush(). However, context switch happens at this time and Thread 2 the cleanup thread set the `AtomicBoolean` to true, and `dataChannel` to `null`. Then context switch back to Thread 1, `dataChannel.flush()` will still hit NPE.

Currently there is no other codes set the `dataChannel`, `indexFile` and `metaFile` to null. And they are created from the file paths which is not expected to be null either. If we really want to distinguish the NPE from the cleanup or other buggy code before cleanup, we can still follow suggestions from @mridulm and also add the AtomicBoolean flag to check whether the dataChannel null is caused by cleanup or not. It will only throw IllegalStatement exception with error message indicating cleanup has been done, if the null is caused by cleanup. Otherwise, NPE will still be thrown out which indicates something else is buggy.

Thoughts? @Ngone51

Reply:
> As discussed offline, let us not do this - this is inherently buggy approach.
> Let us do the alternative discussed -
> 
> a) Remove use of direct field reference, and rely on accessor methods.
> b) In the getter, do the relevant checks to ensure non-null values (and throw exception as relevant).
> c) Use `AtomicReference` for the field related fields in partition info - given they are getting mutated async.
> d) Ensure close/cleanup relies on atomic getAndSet to ensure no MT-safety issues : note, we need to do this only in `AppShufflePartitionInfo`.
> e) We can also make `channel` and `dos` `final` in `MergeShuffleFile` - since `AppShufflePartitionInfo` will atomically handle state transitions.
> 
> (e) is something we had not discussed .. thoughts @zhouyejoe, @otterc ?
> 
> +CC @Ngone51

Thanks for the suggestions. SGTM. Will update accordingly.

Reply:
Double checked on the implementation for the `closeAllFiles` in `AppShufflePartitionInfo`. The major reason of the NPE is the `closeAllFiles` set variables to null. But I don't think we even need to set `dataChannel`, `indexFile` and `metaFile` to null, close() should be sufficient. And the exception from Thread 1, if thread 2 is closing the channels, would be `ClosedChannelException` or just 'IOException'. Similarly, if we make `channel` and `dos` `final` in `MergeShuffleFile`, there is no need to set these two variables to null either. @mridulm @otterc @Ngone51 Thoughts?

Reply:
Just `close` sounds better to me. I'd prefer less change here.

Reply:
Updated with just close, also updated the unit tests.

Reply:
> But I don't think we even need to set dataChannel, indexFile and metaFile to null, close() should be sufficient.
 
 This will raise exceptions (`ClosedChannelException`) when jvm invokes `finalize()` on partitionInfo instance. 
 You will need to make changes to the `finalize` and catch these exceptions per file.

Reply:
We could check for `isOpen` before closing as well.

Reply:
Added check `isOpen`